quil-rs/src/program/source_map.rs

```diff
@@ -0,0 +1,47 @@
+/// A SourceMap provides information necessary to understand which parts of a target
+/// were derived from which parts of a source artifact, in such a way that they can be
+/// mapped in either direction.
+///
+/// The behavior of such mappings depends on the implementations of the generic `Index` types,
+/// but this may be a many-to-many mapping, where one element of the source is mapped (contributed)
+/// to zero or many elements of the target, and vice versa.
+///
+/// This is also intended to be mergeable in a chain, such that the combined result of a series
+/// of transformations can be expressed within a single source mapping.
+#[derive(Clone, Debug, PartialEq)]
+pub struct SourceMap<SourceIndex, TargetIndex> {
```

Comment:
yeah, I've had that in mind from the start, but the first attempt was clumsy given the generics involved. I've added one now, though - good interface, although perhaps less efficient than it could be.

> (maybe, at request) An ergonomic reverse-mapping to allow looking up the source instruction that resulted in a given target instruction. That can always be built by the user, but it might be nice to have one here since the source mapping structure takes a moment to grok even for an experienced Quil dev.